### DISCUSSION TITLE
Autocomplete — share your use cases and implementations

### DISCUSSION THREAD

**Discussion by christopherthomasdesign** [alphagov/govuk-design-system#2374] [Prioritised components and patterns]
We plan to revisit 'Autocomplete' to see if we can make this component more easily available, as well as up-to-date with the latest technologies and accessibility requirements.

## About this work

Autocomplete is a text input that suggests options to the user as they type.

The community has worked on this in the past – see the [accessible autocomplete](https://github.com/alphagov/accessible-autocomplete) repository, and the associated [accessible autocomplete examples](https://alphagov.github.io/accessible-autocomplete/examples/) demo.

Previous conversation can be found at https://github.com/alphagov/govuk-design-system-backlog/issues/4.

## What we're looking for

We want to understand:
- services that use the current autocomplete
- their different use cases
- customisations or ‘hacks’ that teams have made
- datasets in use
- accessibility work service teams have done, and what needs they’ve discovered

**Comment by edwardhorsford** [comment]
Our team has coded an extension to the autocomplete to fix some common usability issues and extend the functionality. It's written as a drop in bit of code other teams can use and then expand on. Most of the features can be configured by adding the appropriate data-attributes on to each item on a select.

[We documented it in this blog post](https://bat-design-history.netlify.app/register-trainee-teachers/autocomplete-improvements/)

Improvements it makes:
* Better matching / searching - whitespace / punctuation differences no longer cause a failed search
* Results are sorted from better match to worse match
* Built-in support for synonyms
* Support for boosting specific items
* Support for stripping stop words (in, of, and, at, etc)
* Support for appending extra information to the name
* Support for showing a hint with the name

We also did some work to have [analytics on our autocompletes](https://bat-design-history.netlify.app/register-trainee-teachers/autocomplete-analytics/) - so we know what people search for and what they ultimately pick - this helped us find gaps in the functionally or where we needed more synonyms.

**Comment by edwardhorsford** [comment]
Our most complex autocomplete is for searching for a school. This uses Ajax to live search our db using the user's query. There are around 40k entries. We use a hint to show additional school details that users can pick from. For non-js, we have a regular input with server-side search.

![schools-autocomplete-name](https://user-images.githubusercontent.com/2204224/194906747-fa5564bf-f058-4c73-9ac0-e6b3974892ba.gif)

[Blog post about the js search](https://bat-design-history.netlify.app/register-trainee-teachers/searching-for-a-school/)
[Blog post about the no-js fallback.](https://bat-design-history.netlify.app/register-trainee-teachers/searching-for-a-school-part-2/)

## Prototyping with the autocomplete

For prototyping, I've wrapped the accessible autocomplete as a nunjucks component. The API could do with a bit of cleanup, but the object currently looks like this:

```nunjucks
{{ appAutocompleteFromSelect({
  label: {
    text: "Type of degree",
    classes: "govuk-label--s"
  },
  hint: {
    text: "Hint"
  },
  id: 'degreeTypeUK',
  name: "degreeTemp[typeUK]",
  items: items,
  classes: "govuk-!-width-two-thirds",
  value: degreeTemp.type,
  autocompleteOptions: {
    minLength: 2,
    autoselect: false,
    showSuggestionsBanner: true,
    values: degreeTypeAutocompleteValues // optional values
  }
} ) }}
```

The component then renders out an html select with appropriate data attributes on each option. Our autocomplete code runs on all pages and inits for each select or input it finds, and gets the relevant data from the data attributes. This component has made setup of new autocompletes for prototypes much easier - as you usually just need to provide an array of data. There's a similar component for enhancing from an input rather than a select.

**Reply by rgarner** [reply]
@edwardhorsford we have the same use case, with the same dataset/size. What we've found is that the autocomplete (even with `minLength`) generates a lot of queries, which puts demand/expense on hosting for smaller projects (typically those relying on commodity Postgres as opposed to some search index). Debounce helps here https://github.com/alphagov/accessible-autocomplete/pull/611.

**Comment by edwardhorsford** [comment]
DfE has also been working on an autocomplete with revised styling for situations where the results are suggestions only, but the user is free to type whatever they want. This work was done with @frankieroberto and was influenced by the ONS' Census 2021 autocomplete.

Autocomplete with suggestions:
![autocomplete-with-suggestions3](https://user-images.githubusercontent.com/2204224/194907529-ecbf2e6c-20b7-4207-9c4f-168e556a2683.gif)

ONS Census 2021 autocomplete:
![census-country-of-birth-2](https://user-images.githubusercontent.com/2204224/194907906-c4834bbe-433e-4952-8ec1-09c90ab2c009.gif)

**Reply by christopherthomasdesign** [reply]
These are very nice indeed 😎

**Comment by edwardhorsford** [comment]
One area where the autocomplete lacks is proper support for values that are separate from the visible text. Eg the text might be "Hogwarts academy" but you want to store a uuid value. This becomes important for larger sets where you might have several items with the same name.

Selects support values and it half-works if you're using the `enhanceSelectElement` mode - but not very well, and not if you're enhancing from an input. We've patched this by creating a hidden input that stores the value - and sending that to the server.

**Comment by adamliptrot-oc** [comment]
Working on service teams at HMRC I added [some js and css to handle a few issues (link to Gist):](https://gist.github.com/adamliptrot-oc/f7cbb92f040082cc17cff27416ae348b)

1. a user picking a valid answer, not submitting it, changing their answer to an invalid one and submitting - the previously selected valid response was sent instead of the new invalid one.
2. error styling not being presented on the js-generated input
3. any aria-describedby attributes on the original select not being copied over to the js-generated input (so any hints or errors are not exposed as expected)
4. css to handle the 'show all' icon not presenting due to z-index

**Comment by DilwoarH** [comment]
On GOVUK Publishing we have a need for multiple selections as well as a single selection. An example of this can be found [here](https://content-publisher.integration.publishing.service.gov.uk/component-guide/autocomplete#select_multiple).

From my experience in working on publishing apps, I found that the following is needed to make the accessible autocomplete better:
1. If you select an option, it should not allow you to edit this text (unless you want to select/search for another option).
2. It should allow multiple items to be selected.
3. It should only allow you to select from the list of items specified - ie. You shouldn't be allowed to enter text which isn't in the list.
4. It should allow you to use [select option groups](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/optgroup).
5. It should allow users to scroll through the list without requiring them to enter a letter.
6. For long lists, it should support AJAX requests so that the DOM doesn't have to load up 1000s of options.

The above basically describes the behaviours of the [select2 plugin](https://select2.org/) which is slightly different to the current implementation of the GOVUK Accessible Autocomplete. In my **opinion**, an accessible "select2" type component is what we need for almost all cases.

**Comment by edwardhorsford** [comment]
I feel like it would be worth exploring multiple standalone components serving different needs. Or at least thinking about the various use-cases and how our guidance could document them.

Part of the challenge of the current autocomplete is that it can be extended very flexibly, but it's up to teams to modify, and you basically have to customise everything to get it working well and fix the usability gotchas. Different settings can have usability implications, but it's not obvious what these are. It would be so helpful to have worked examples of the different ways you can set it up, and the pros and cons of them.

Some possible components:

## Enhanced select
* To help users pick from a smallish number of fixed options
* Fix many of the usability issues with selects
* Very limited in configurability - but easy for teams to implement - just provide an array.
* Possibly styled closer to a drop-down/ select rather than an input.

## Suggestion autocomplete
* Similar to [examples above](https://github.com/alphagov/govuk-design-system/discussions/2374#discussioncomment-3840815) 
* Uses results as suggestions, but users are free to type their own thing

## Autocomplete advanced
* Possibly a bit similar to current autocomplete, but likely rewritten
* For searching from larger lists
* Users normally expected to pick from list
* Configurable for the various use cases teams have
* Sensible defaults / presets for common use cases.
* Wrapped as a Nunjucks component but can be configured with js if needed

----

**Comment by frankieroberto** [comment]
We use autocompletes in a few places within [Apply for teacher training](http://www.apply-for-teacher-training.service.gov.uk/).

They fall into 2 categories.

## 1) Where users have to select an item from the list

These enhance a `<select class="govuk-select">` element, which is shown if javascript is unavailable.

### Nationality
<img width="718" alt="Screenshot 2022-10-21 at 13 11 32" src="https://user-images.githubusercontent.com/30665/197192818-13b0da92-1a66-4440-b85c-40f28c4c1673.png">

### Training provider
<img width="710" alt="Screenshot 2022-10-21 at 13 12 21" src="https://user-images.githubusercontent.com/30665/197192958-edc6e0e5-029c-4053-a688-fd92babdf1e2.png">

## 2) Where there is a list, but users can also type in their own answer

These enhance a `<input type="text" class="govuk-input">` element.

### Degree subject

<img width="692" alt="Screenshot 2022-10-21 at 13 16 30" src="https://user-images.githubusercontent.com/30665/197193774-23becc07-9352-4630-97a5-97e333190e0a.png">

### University

<img width="694" alt="Screenshot 2022-10-21 at 13 17 43" src="https://user-images.githubusercontent.com/30665/197193840-69798172-1680-45b4-a923-3b14813581f9.png">

## What we’ve found

It hasn’t always been clear whether users _have_ to pick an option from the list or not. This could lead to users picking the "closest match" answer rather than typing in an accurate answer. To resolve this we added hint text, a "Suggestions" header and a gap between the input box and the options list. When there are no matches, we do not display the list or the heading, so that the input looks like a regular text input. In user research, we observed participants being able to both select an option and typing in their own answers.

Where a user has to pick an item from the list, we generally display a down arrow, and have the list appear when focusing the input even if no characters have been entered. This makes it more similar to a regular dropdown list, so that users realise they have to pick an option. The downside to this is that not all users realise they can type in some characters to filter the list.

**Reply by christopherthomasdesign** [reply]
Cheers Frankie! I'd be interested to hear a bit more about the need behind option 2, where a user can type what they like but we give them suggestions as they type. Is it because users don't know what to type? Or does it help with matching data in the backend somewhere?

**Reply by frankieroberto** [reply]
@christopherthomasdesign hiya. sorry for slow reply. The main benefit for the user of the option 2, where the autocomplete is added to a free text field, is that it can speed up their task and give them confidence that they have entered information in correctly. For example, when typing in their degree subject or university they might see their answer after only typing in a few characters, which saves a (small amount) of time and gives them a satisfying feeling that their answer was known and that they haven't made any typos, spelling mistakes or used a different term.

It's a relatively marginal benefit for the user (depending on how long the overall task is), but I think this is one reason why it's acceptable as an "enhancement", with the no-javascript option of just having a free text field as an adequate fallback.

As a secondary benefit, the autocomplete also gives the service "cleaner" data, with less likelihood of duplicates caused by typos, alternative names, and so on. This shouldn't be the only reason for doing it (as you could clean up that data separately), but can be an important side benefit when doing analysis of the data.

**Comment by edwardhorsford** [comment]
My team previously used the `showAllValues` option that displays the arrow and shows everything on focus. We ultimately removed it because of several usability issues.

Issues:
* Users *never* searched / typed when this mode was used - they just scrolled the list. This may be good for short lists, but for long lists 95% of the value of the autocomplete is lost. This is especially true if you want users to be able to search with synonyms.
* Users expect to be able to click the arrow to close the list. This either doesn't work or is buggy in the autocomplete.
* Once an answer is selected, users expect to be able to click the arrow to see all options again - the autocomplete does not do this.

I think having such a mode is worth keeping (see enhanced select suggestion above) - but would have it as a standalone thing and note that it's more suited to lists with not many items. I'd also restyle explicitly like a menu rather than an input.

**Reply by rgarner** [reply]
@edwardhorsford We're about to do this for subject specialisms. The insights are appreciated.

**Reply by katydelamotte** [reply]
Hi @edwardhorsford @rgarner did you put any rules in place to limit a potentially very long list (i.e., minimum number of characters typed), or have any insights about users navigating long lists from autocomplete? I'm looking at using autocomplete for a very large data set in HMCTS and considering a 'show all', so this has been useful to read (even if a couple of years old!) 
Thanks :)

**Reply by rgarner** [reply]
@katydelamotte sorry Katy just noticed this (I know...) yes we did – minimum number of characters typed. No real insights as our largest list was about 40K and the min characters (defaulting to 3) and a max showable page size of 50 didn't throw up any problems.

**Comment by edwardhorsford** [comment]
I think the autocomplete could also do with in-built support for a 'clear' button or link. It's a PITA to use where there's a pre-filled value - it's not obvious you have to clear all text to search again.

**Comment by edwardhorsford** [comment]
There's [this 2019 blog post](https://accessibility.blog.gov.uk/2019/04/08/accessibility-lessons-dealing-with-a-large-amount-of-form-inputs/) from the GDS blog post investigating selecting multiple items for GOV.UK Finders that covers autocompletes.

In particular they looked at some ways of possibly supporting selecting multiple:
![image4](https://user-images.githubusercontent.com/2204224/199277026-728b77cd-bfb0-4118-919b-9e11030b11ec.png)

**Reply by stevenjmesser** [reply]
I remember this work evolving while I was working on the Search team, and the current version is (I think) on the super-group finders, e.g. [Guidance and regulation](https://www.gov.uk/search/guidance-and-regulation). 

<img width="303" alt="image" src="https://user-images.githubusercontent.com/23193154/208714472-2695dc63-5c4d-4cbe-b043-98a3669ee4e0.png">

**Comment by jessdenney** [comment]
**HMRC Online Payments Service**

Part of our open banking journey requires the user to pick out the name of their bank account from a list.

Its quite a long list and it was originally trialed with radios but it wasn't very practical so we now use the auto complete list. This works quite well as there are some items which are very similar such as Bank of Scotland Personal and Bank of Scotland Business.

I'm not sure if any 'hacks' have been used to improve the function for our users. But if their bank isn't listed because its not part of open banking the user is shown other payment options behind the secondary button 'my account is not listed'.

![Screenshot 2022-11-14 at 16 24 15](https://user-images.githubusercontent.com/73240363/201712398-5bd9a07d-e348-4be1-bf44-633d33133650.png)

**Reply by jessdenney** [reply]
I thought it would be good to share how we updated this design a while ago to consider javascript enabled/disabled. We did this as some accessibility software disables javascript.

Javascript enabled (default version)
<img width="1016" alt="Screenshot 2025-01-16 at 09 52 54" src="https://github.com/user-attachments/assets/c5ac9194-fb7e-4804-86b4-42dc1e589d60" />

Javascript disabled - autocomplete will not work and select component shown instead
<img width="996" alt="Screenshot 2025-01-16 at 09 53 26" src="https://github.com/user-attachments/assets/18ea507e-69f1-465a-84cb-5ce4559583cb" />

**Comment by adyhoran1** [comment]
We use the autocomplete component to assign projects to internal Department for Education staff. See https://conversions-design-history.netlify.app/apply-to-become-academy/assign-a-project/#prototype-design

We've noticed the following issue:

Select a user from the autocomplete list

![image](https://user-images.githubusercontent.com/97228674/203577574-4c83b2ac-4355-437c-87eb-6d7e07a7ae16.png)

Change your mind and delete the contents of the auto-complete box

![Empty](https://user-images.githubusercontent.com/97228674/203577777-993f1d39-65d0-4ed4-97ab-3308e8aa3c45.png)

Click Continue. The person that was selected but deleted is assigned

<img width="830" alt="Assigned" src="https://user-images.githubusercontent.com/97228674/203578464-c2707513-6890-4d42-8518-029b9f22d4ee.png">

We don't think this should happen - we think if you delete the text in the text field and click Continue, then it should not assign someone to the project.

**Reply by adamliptrot-oc** [reply]
@adyhoran1 We have seen this too - see earlier comment https://github.com/alphagov/govuk-design-system/discussions/2374#discussioncomment-3879635

**Reply by edwardhorsford** [reply]
@adyhoran1 it's possible for you to fix / correct this. You can set up the autocomplete to submit the raw input string as well as the 'selected' select item. Your backend can then compare the two or see if it's blank and behave appropriately.

I think in our prototype I also patched this in js so that we clear the select too if the value in the input is blank.

**Comment by martinwake** [comment]
A possible use case for what's called Advanced Autocomplete above is address lookups. See for example the Royal Mail's at https://www.royalmail.com/find-a-postcode.

**Comment by CharlotteDowns** [comment]
We've just [added some guidance on how to 'share findings about your users' with us](https://design-system.service.gov.uk/community/share-research-findings/) 📝. This will help us learn more about how your users autocomplete within your service.

**Comment by armoj** [comment]
I submitted [this question/bug](https://github.com/alphagov/accessible-autocomplete/issues/605) to your accessible-autocomplete issues a few days ago. There is a problem with our autocomplete when used with Dragon NaturallySpeaking. There is an issue, either with the autocomplete, or Dragon that prevents the form value being updated. I was wondering if this is a bug/issue that you are aware of at this time.

**Comment by stephenjmcneill1971** [comment]
Is there anyway to make the assistiveHint attribute optional?

I would like to removde the assistiveHint text

"When autocomplete results are available use up and down arrows to review and enter to select. Touch device users, explore by touch or with swipe gestures."

and the related aria-describedby being rendered. Is this an option.

Alternatively could we make it visible and styled

**Reply by edwardhorsford** [reply]
One of the options lets you provide your own assistiveHint - I assume this could be used to effectively 'clear' it. Though it wouldn't be recommended, because it's probably good to tell screen reader users how to interact with the component.

If you want instead for something to be visually shown, you'd need to add this yourself - eg use javascript to insert the appropriate element in the DOM after the component has loaded.

**Reply by edwardhorsford** [reply]
Note: I raised a [bug earlier this year about AssistiveHints being broken in 2.04](https://github.com/alphagov/accessible-autocomplete/issues/590).

**Comment by ryanb-gds** [comment]
To add to [the comment above](https://github.com/alphagov/govuk-design-system/discussions/2374#discussioncomment-3881123),  GOV.UK Publishing are still currently maintaining a version of the autocomplete which allows for multiple selections at https://github.com/alphagov/accessible-autocomplete-multiselect.

However, it has recently failed an accessibility review because it does not announce the selections effectively or the presence of the "Remove" button for screenreaders. We are yet to decide at this point in time whether or not to invest in making the existing multiselect implementation more accessible, which we expect will be challenging, or to explore an alternative design pattern.

**Comment by JillRichardsonPratt** [comment]
Would it be possible to call this component something other than 'autocomplete' so that we aren't using the same name for both this component and the autocomplete attribute that helps users to reuse previously filled data? When working on a team that has tickets relating to the autocomplete attribute alongside other tickets relating to this component, it's tricky to work out which tickets refer to which things! It would be great it we could remove the ambiguity. :)

**Comment by clare-brown** [comment]
**HMRC Economic Crime Levy** 

We've used this component successfully as part of our registration journey where we ask users 'Who is your AML supervisor?' - but deviating from the standard component using it within a radio dropdown.

It was done to create focus on the main 3 supervisors, HMRC, the Gambling Commission and the Financial Conduct Authority and reduce the amount of 'Other' options on screen as it's quite a long list of 22 other professional bodies. 

It's tested well after iterating and adding 'Name of professional body' above the hint text.

<img width="827" alt="image-2024-4-5_13-44-2" src="https://github.com/alphagov/govuk-design-system/assets/107033520/1c18a4dc-1bf5-4bce-95b0-4dd14820240c">

**Comment by james-work-account** [comment]
Hi all, just sharing an issue which at least one of our live users has had with the accessible-autocomplete component, specifically how it is pre-filled when going back to change their answer after initially entering something they didn't mean to.

> We've seen a couple of users having issues with accessible-autocomplete, namely that if they select an option then go back to the page they don't know that they need to clear their previous answer in order for others to show up. i.e. because the accessible-autocomplete pre-fills the previous answer when they go back to the page, and the accessible-autocomplete is narrowing down the options it shows to exactly what they've selected and nothing else, the user is getting confused as it's not clear to them that they need to clear the field for other options to show up.

I've recorded a video which hopefully clarifies the issue they're having:


https://github.com/alphagov/govuk-design-system/assets/31282758/83d50955-3913-4558-975e-2addeb281a3f

The first half of the video is the issue they're having, the second half is me spitballing a solution (you can ignore the second half - this was initially sent to a teammate who suggested I post on here).

**Reply by edwardhorsford** [reply]
@james-work-account agree this is an issue. I think the ideal would be that if `showAllValues` is enabled, when focusing the field with an existing value, then all results would be shown. Alas this isn't currently supported. You could half hack around it by clearing on focus or adding a clear button, but it's not ideal.

My personal experience has been that the `showAllValues` feature is often more trouble than it's worth - and have tended to avoid using it in services, especially where there are more than a handful of values.

**Reply by james-work-account** [reply]
We've implemented this as a workaround - a button on the page which clears the current input when clicked with JavaScript. It's not perfect but it's been received well by people who've tried it out.

https://github.com/alphagov/govuk-design-system/assets/31282758/3a5ee4e2-edfc-4dfd-b666-ed84dca4bc1c

It was a headache to get to work, so here's the Twirl template in case anyone else needs/wants to implement something similar:

```html
@import models.requests.DataRequest
@import views.html.helper.CSPNonce

@this(govukButton: GovukButton)

@(idToClear: String)(implicit request: DataRequest[_], messages: Messages)

@govukButton(
    ButtonViewModel(Text(messages("site.clearSelectedCode")))
    .withCssClass("no-print")
    .withId("clear-selected-code")
    .withAttribute("type", "button")
    .asSecondaryButton()
)

<script @CSPNonce.attr>
        document.getElementById('clear-selected-code').addEventListener('click', function () {
            const element = document.getElementById('@idToClear');
            element.value = "";
            element.click();
            element.focus();
        });
</script>
```

**Comment by martin-jordan** [comment]
To understand: Is this still the work that @edwardhorsford [had blogged about in spring 2017](https://designnotes.blog.gov.uk/2017/04/20/were-building-an-autocomplete/)? Is that work still ongoing?

**Reply by frankieroberto** [reply]
@martin-jordan I think that work led to the [accessible autocomplete](https://github.com/alphagov/accessible-autocomplete/) but that predates the design system and hasn't yet been incorporated into it or updated to use the same javascript and css framework.

The autocomplete is still on the backlog for the GOV.UK Design System team. I’m not sure where it currently sits in their roadmap or prioritisation.

**Comment by cfgreene** [comment]
Thank you to everyone that has made the [component available](https://github.com/alphagov/accessible-autocomplete/) - we are hoping to use it in our org, we forked from NHS ~3 years ago and when we need new components we check GOV and NHS first. 

I was reading the recent blog post about the [Autocomplete being in use](https://insidegovuk.blog.gov.uk/2024/12/12/making-it-quicker-and-easier-to-search-on-gov-uk/)

Would anyone know when it might be documented in the Design System?

I would have commented and asked via the Blog post but the comment functionality wasn't available. 

Thank you for any info.

**Comment by jamiehylands** [comment]
Defra Livestock Information Service

We have recently been testing the autocomplete component forked from the Digital Land service (https://digital-land.github.io/design-system/components/autocomplete/) as part of our Livestock Registration journey. 

In order to improve data integrity and reduce transposition errors, the field pulls in data from a fixed list of ear tag numbers which are connected to a livestock keeper and their holding. 

Testing has been very successful to date and we are hoping to ingest a version of the component into our in-house pattern library going forward. 

Content design has been critical to the success of this component – our CD shaped the hint text iteratively based on user feedback so keepers feel comfortable adding full and partial ear tag numbers in order to generate results. 

<img width="1199" height="657" alt="CPT2609150907-1199x657" src="https://github.com/user-attachments/assets/5fbd2857-c77f-4d36-8e3e-319f3ff9d081" />
